app/session.server.ts

```diff
@@ -35,7 +40,101 @@ const { getSession, commitSession, destroySession } =
     },
   });
 
-export { getSession, commitSession, destroySession };
+const fiveMinutesInSeconds = 60 * 5;
+const fiveMinutesInMilliseconds = fiveMinutesInSeconds * 1000;
+
+const twoWeeksInSeconds = 60 * 60 * 24 * 14;
+const twoWeeksInMilliseconds = twoWeeksInSeconds * 1000;
+
+// Session last 2 weeks by default. For testing purposes, we can add a
+// DEBUG_SESSION environment variable to make the sessions last 5 minutes, which
+// is the minimum allowed time for a session
+const COOKIE_EXPIRATION = {
+  seconds: process.env.DEBUG_SESSION ? fiveMinutesInSeconds : twoWeeksInSeconds,
+  milliseconds: process.env.DEBUG_SESSION
+    ? fiveMinutesInMilliseconds
+    : twoWeeksInMilliseconds,
+};
+
+/**
+ * Use this to register a new user or log in an existing user. This creates a
+ * session cookie, which can then be used to authenticate further requests.
+ *
+ * This method has 2 potential side effects:
+ * - if the user doesn't exist, it creates a new User document in Firestore
+ * - if the user doesn't have a profile, it creates a new UserProfile document
+ *   in Firestore
+ */
+export async function createUserSession(idToken: string, accessToken: string) {
+  const token = await auth.createSessionCookie(idToken, {
+    expiresIn: COOKIE_EXPIRATION.milliseconds,
+  });
+  const session = await storage.getSession();
+  session.set("token", token);
+
+  // Check that we have a matching user for this token. If not, create one.
+  const decodedToken = await auth.verifySessionCookie(token, true);
+  let user = await getUserByUid(decodedToken.uid);
+
+  if (user == null) {
+    user = await createUser(decodedToken, accessToken);
+  }
+
+  // Check whether the user has a profile page. If not, ask them to create one.
+  let redirectPath: string;
+  const profile = await getUserProfileBySlug(user.githubLogin);
+
+  if (profile) {
+    // If the user already has a profile, we redirect to it
+    redirectPath = `/u/github/${user.githubLogin}`;
+  } else {
+    // If the user doesn't have a profile yet, we create one and send them to
+    // the Welcome page to enter some information
+    await createProfileForUser(user);
+
+    redirectPath = `/u/github/${user.githubLogin}/welcome`;
+  }
+
+  return redirect(redirectPath, {
+    headers: {
+      "Set-Cookie": await storage.commitSession(session),
+    },
+  });
+}
+
+export async function getCurrentSession(request: Request) {
+  const cookieSession = await storage.getSession(request.headers.get("Cookie"));
+  const token = cookieSession.get("token");
+  if (!token) return null;
+
+  try {
+    const tokenUser = await auth.verifySessionCookie(token, true);
```

Comment:
The second big change is that we now verify the session cookie, instead of verifying the `idToken`.